File: Library/src/main/java/org/endeavourhealth/library/model/iml/IMLLanguage.java
package org.endeavourhealth.library.model.iml;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import lombok.Getter;

public class IMLLanguage {

  @Getter
  private String text;

  @Getter
  private String lang;

  @Getter
  private Map<String, String> info = new HashMap<>();

  private Map<String, String> definitions = new HashMap<>();
  private Map<String, String> prefixes = new HashMap<>();

  @Getter
  private Set<String> keywords = new HashSet<>();

  @Getter
  Set<String> booleans = new HashSet<>();

  @Getter
  Set<String> alerts = new HashSet<>();

  private Map<String, Set<String>> iriVariables = new HashMap<>();

  public Map<String, Set<String>> getIriVariables() {
    return iriVariables;
  }

  public IMLLanguage() {
    this.getKeywords().addAll(Set.of("define", "assign", "match", "as", "from", "where", "if", "prefix", "info", "default"));
    this.getBooleans().addAll(Set.of("either", "and", "or"));
    this.getAlerts().addAll(Set.of("exclude", "warning"));
  }

  public IMLLanguage setIriVariables(Map<String, Set<String>> iriVariables) {
    this.iriVariables = iriVariables;
    return this;
  }

  public IMLLanguage setText(String text) {
    this.text = text;
    return this;
  }

  public IMLLanguage setLang(String lang) {
    this.lang = lang;
    return this;
  }

  public IMLLanguage setInfo(Map<String, String> info) {
    this.info = info;
    return this;
  }

  public Map<String, String> getDefinitions() {
    return definitions;
  }

  public IMLLanguage setDefinitions(Map<String, String> definitions) {
    this.definitions = definitions;
    return this;
  }

  public IMLLanguage setKeywords(Set<String> keywords) {
    this.keywords = keywords;
    return this;
  }

  public Map<String, String> getPrefixes() {
    return prefixes;
  }

  public IMLLanguage setPrefixes(Map<String, String> prefixes) {
    this.prefixes = prefixes;
    return this;
  }
}

File: src/models/IMLLanguage.ts
import z from "zod";

// export interface IMLLanguage {
//   text?: string;
//   lang?: string;
//   info?: { [index: string]: string };
//   definitions?: { [index: string]: string };
//   prefixes?: { [index: string]: string };
//   keywords?: string[];
//   booleans?: string[];
//   alerts?: string[];
//   iriVariables?: { [index: string]: string[] };
// }

export const IMLLanguageSchema = z.strictObject({
  text: z.string().optional(),
  lang: z.string().optional(),
  info: z.record(z.string(), z.string()).optional(),
  definitions: z.record(z.string(), z.string()).optional(),
  prefixes: z.record(z.string(), z.string()).optional(),
  keywords: z.array(z.string()).prefault([]),
  booleans: z.array(z.string()).prefault([]),
  alerts: z.array(z.string()).prefault([]),
  iriVariables: z.record(z.string(), z.array(z.string()).prefault([])).optional()
});

export type IMLLanguage = z.output<typeof IMLLanguageSchema>;

export function isIMLLanguage(value: unknown): value is IMLLanguage {
  return IMLLanguageSchema.safeParse(value).success;
}
